File: Co-Simulation/Sumo/sumo-1.7.0/tools/contributed/traas/src/main/java/de/tudresden/sumo/cmd/Inductionloop.java
/****************************************************************************/
// Eclipse SUMO, Simulation of Urban MObility; see https://eclipse.org/sumo
// Copyright (C) 2017-2020 German Aerospace Center (DLR) and others.
// TraaS module
// Copyright (C) 2016-2017 Dresden University of Technology
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License 2.0 which is available at
// https://www.eclipse.org/legal/epl-2.0/
// This Source Code may also be made available under the following Secondary
// Licenses when the conditions for such availability set forth in the Eclipse
// Public License 2.0 are satisfied: GNU General Public License, version 2
// or later which is available at
// https://www.gnu.org/licenses/old-licenses/gpl-2.0-standalone.html
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-or-later
/****************************************************************************/
/// @file    Inductionloop.java
/// @author  Mario Krumnow
/// @author  Evamarie Wiessner
/// @date    2016
///
//
/****************************************************************************/
package de.tudresden.sumo.cmd;
import de.tudresden.sumo.config.Constants;
import de.tudresden.sumo.util.SumoCommand;

/**
 *
 * @author Mario Krumnow
 * @author Evamarie Wiessner
 *
 */

public class Inductionloop {

    //getter methods

    /**
     * Returns a list of IDs of all induction loops within the scenario.
     *
     * @return list of IDs of all induction loops in the network
     */

    public static SumoCommand getIDList() {
        return new SumoCommand(Constants.CMD_GET_INDUCTIONLOOP_VARIABLE, Constants.TRACI_ID_LIST, "", Constants.RESPONSE_GET_INDUCTIONLOOP_VARIABLE, Constants.TYPE_STRINGLIST);
    }


    /**
     * Returns the number of all induction loops in the network.
     * @return the number of detectors
     */

    public static SumoCommand getIDCount() {
        return new SumoCommand(Constants.CMD_GET_INDUCTIONLOOP_VARIABLE, Constants.ID_COUNT, "", Constants.RESPONSE_GET_INDUCTIONLOOP_VARIABLE, Constants.TYPE_INTEGER);
    }

    /**
     * Returns the chosen parameter
     *
     * @param personID a string personIDentifying the person
     *  @param param a string personIDentifying the parameter
     *
     * @return the specific parameter
     */

    public static SumoCommand getParameter(String loopID, String param) {
        Object[] array = new Object[] {param};
        return new SumoCommand(Constants.CMD_GET_INDUCTIONLOOP_VARIABLE, Constants.VAR_PARAMETER, loopID, array, Constants.RESPONSE_GET_INDUCTIONLOOP_VARIABLE, Constants.TYPE_STRING);
    }


    /**
     * Returns the ID of the lane the loop is placed at.
     *
     * @param loopID
     *            a string identifying the induction loop
     * @return a string identifying the lane
     */

    public static SumoCommand getLaneID(String loopID) {
        return new SumoCommand(Constants.CMD_GET_INDUCTIONLOOP_VARIABLE, Constants.VAR_LANE_ID, loopID, Constants.RESPONSE_GET_INDUCTIONLOOP_VARIABLE, Constants.TYPE_STRING);
    }

    /**
     * Returns the mean length of vehicles (in m) which were on the detector in
     * the last step.
     *
     * @param loopID
     *            a string identifying the induction loop
     * @return average length of the vehicles (in m)
     */

    public static SumoCommand getLastStepMeanLength(String loopID) {
        return new SumoCommand(Constants.CMD_GET_INDUCTIONLOOP_VARIABLE, Constants.LAST_STEP_LENGTH, loopID, Constants.RESPONSE_GET_INDUCTIONLOOP_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns the mean speed of vehicles (in m/s) that were on the named
     * induction loop within the last simulation step.
     *
     * @param loopID
     *            a string identifying the induction loop
     * @return average speed of the vehicles (in m/s)
     */

    public static SumoCommand getLastStepMeanSpeed(String loopID) {
        return new SumoCommand(Constants.CMD_GET_INDUCTIONLOOP_VARIABLE, Constants.LAST_STEP_MEAN_SPEED, loopID, Constants.RESPONSE_GET_INDUCTIONLOOP_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns the percentage of time (in %) the detector was occupied by a
     * vehicle within the last time step.
     *
     * @param loopID
     *            a string identifying the induction loop
     * @return occupancy of the detector (in %)
     */

    public static SumoCommand getLastStepOccupancy(String loopID) {
        return new SumoCommand(Constants.CMD_GET_INDUCTIONLOOP_VARIABLE, Constants.LAST_STEP_OCCUPANCY, loopID, Constants.RESPONSE_GET_INDUCTIONLOOP_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Returns the list of IDs of vehicles that were on the named induction loop
     * in the last simulation step.
     *
     * @param loopID
     *            a string identifying the induction loop
     * @return list of all vehicleIDs
     */

    public static SumoCommand getLastStepVehicleIDs(String loopID) {
        return new SumoCommand(Constants.CMD_GET_INDUCTIONLOOP_VARIABLE, Constants.LAST_STEP_VEHICLE_ID_LIST, loopID, Constants.RESPONSE_GET_INDUCTIONLOOP_VARIABLE, Constants.TYPE_STRINGLIST);
    }

    /**
     * Returns the number of vehicles that were on the named induction loop
     * within the last simulation step.
     *
     * @param loopID
     *            a string identifying the induction loop
     * @return total number of vehicles
     */

    public static SumoCommand getLastStepVehicleNumber(String loopID) {
        return new SumoCommand(Constants.CMD_GET_INDUCTIONLOOP_VARIABLE, Constants.LAST_STEP_VEHICLE_NUMBER, loopID, Constants.RESPONSE_GET_INDUCTIONLOOP_VARIABLE, Constants.TYPE_INTEGER);
    }

    /**
     * Returns the position (in m) of the induction loop at it's lane, counted
     * from the lane's begin.
     *
     * @param loopID
     *            a string identifying the induction loop
     * @return position of the detector, counted from the lane's begin
     */

    public static SumoCommand getPosition(String loopID) {
        return new SumoCommand(Constants.CMD_GET_INDUCTIONLOOP_VARIABLE, Constants.VAR_POSITION, loopID, Constants.RESPONSE_GET_INDUCTIONLOOP_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * Return the time (in s) since last detection.
     *
     * @param loopID
     *            a string identifying the induction loop
     * @return time since last detection (in s)
     */

    public static SumoCommand getTimeSinceDetection(String loopID) {
        return new SumoCommand(Constants.CMD_GET_INDUCTIONLOOP_VARIABLE, Constants.LAST_STEP_TIME_SINCE_DETECTION, loopID, Constants.RESPONSE_GET_INDUCTIONLOOP_VARIABLE, Constants.TYPE_DOUBLE);
    }

    /**
     * A complex structure containing several information about vehicles which
     * passed the detector.
     *
     * @param loopID
     *            a SumoStringList containing induction loops
     * @return several information about vehicles which passed the detector.
     */

    public static SumoCommand getVehicleData(String loopID) {
        return new SumoCommand(Constants.CMD_GET_INDUCTIONLOOP_VARIABLE, Constants.LAST_STEP_VEHICLE_DATA, loopID, Constants.RESPONSE_GET_INDUCTIONLOOP_VARIABLE, Constants.TYPE_COMPOUND);
    }

    //setter methods


}

File: Co-Simulation/Sumo/sumo-1.7.0/tools/traci/_inductionloop.py
# -*- coding: utf-8 -*-
# Eclipse SUMO, Simulation of Urban MObility; see https://eclipse.org/sumo
# Copyright (C) 2011-2020 German Aerospace Center (DLR) and others.
# This program and the accompanying materials are made available under the
# terms of the Eclipse Public License 2.0 which is available at
# https://www.eclipse.org/legal/epl-2.0/
# This Source Code may also be made available under the following Secondary
# Licenses when the conditions for such availability set forth in the Eclipse
# Public License 2.0 are satisfied: GNU General Public License, version 2
# or later which is available at
# https://www.gnu.org/licenses/old-licenses/gpl-2.0-standalone.html
# SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-or-later

# @file    _inductionloop.py
# @author  Michael Behrisch
# @author  Daniel Krajzewicz
# @date    2011-03-16

from __future__ import absolute_import
from .domain import Domain
from . import constants as tc


def readVehicleData(result):
    result.readLength()
    nbData = result.readInt()
    data = []
    for _ in range(nbData):
        vehID = result.readTypedString()
        length = result.readTypedDouble()
        entryTime = result.readTypedDouble()
        leaveTime = result.readTypedDouble()
        typeID = result.readTypedString()
        data.append((vehID, length, entryTime, leaveTime, typeID))
    return data


_RETURN_VALUE_FUNC = {tc.LAST_STEP_VEHICLE_DATA: readVehicleData}


class InductionLoopDomain(Domain):

    def __init__(self):
        Domain.__init__(self, "inductionloop", tc.CMD_GET_INDUCTIONLOOP_VARIABLE, None,
                        tc.CMD_SUBSCRIBE_INDUCTIONLOOP_VARIABLE, tc.RESPONSE_SUBSCRIBE_INDUCTIONLOOP_VARIABLE,
                        tc.CMD_SUBSCRIBE_INDUCTIONLOOP_CONTEXT, tc.RESPONSE_SUBSCRIBE_INDUCTIONLOOP_CONTEXT,
                        _RETURN_VALUE_FUNC,
                        subscriptionDefault=(tc.LAST_STEP_VEHICLE_NUMBER,))

    def getPosition(self, loopID):
        """getPosition(string) -> double

        Returns the position measured from the beginning of the lane in meters.
        """
        return self._getUniversal(tc.VAR_POSITION, loopID)

    def getLaneID(self, loopID):
        """getLaneID(string) -> string

        Returns the id of the lane the loop is on.
        """
        return self._getUniversal(tc.VAR_LANE_ID, loopID)

    def getLastStepVehicleNumber(self, loopID):
        """getLastStepVehicleNumber(string) -> integer

        Returns the number of vehicles that were on the named induction loop within the last simulation step.
        """
        return self._getUniversal(tc.LAST_STEP_VEHICLE_NUMBER, loopID)

    def getLastStepMeanSpeed(self, loopID):
        """getLastStepMeanSpeed(string) -> double

        Returns the mean speed in m/s of vehicles that were on the named induction loop within the last simulation step.
        """
        return self._getUniversal(tc.LAST_STEP_MEAN_SPEED, loopID)

    def getLastStepVehicleIDs(self, loopID):
        """getLastStepVehicleIDs(string) -> list(string)

        Returns the list of ids of vehicles that were on the named induction loop in the last simulation step.
        """
        return self._getUniversal(tc.LAST_STEP_VEHICLE_ID_LIST, loopID)

    def getLastStepOccupancy(self, loopID):
        """getLastStepOccupancy(string) -> double

        Returns the percentage of time the detector was occupied by a vehicle.
        """
        return self._getUniversal(tc.LAST_STEP_OCCUPANCY, loopID)

    def getLastStepMeanLength(self, loopID):
        """getLastStepMeanLength(string) -> double

        Returns the mean length in m of vehicles which were on the detector in the last step.
        """
        return self._getUniversal(tc.LAST_STEP_LENGTH, loopID)

    def getTimeSinceDetection(self, loopID):
        """getTimeSinceDetection(string) -> double

        Returns the time in s since last detection.
        """
        return self._getUniversal(tc.LAST_STEP_TIME_SINCE_DETECTION, loopID)

    def getVehicleData(self, loopID):
        """getVehicleData(string) -> [(veh_id, veh_length, entry_time, exit_time, vType), ...]

        Returns a complex structure containing several information about vehicles which passed the detector.
        """
        return self._getUniversal(tc.LAST_STEP_VEHICLE_DATA, loopID)
